deepspeed/runtime/engine.py

```diff
@@ -2899,6 +2901,20 @@ def _load_zero_checkpoint(self, load_dir, tag, load_optimizer_states=True):
             logger.info(f"loading {len(zero_sd_list)} zero partition checkpoints for rank {self.global_rank}")
         return True
 
+    def update_optimizer_step(self, step):
+        optimizer = self.optimizer
+        base_optimizer = optimizer.optimizer
+        state = base_optimizer.state
+        for group in optimizer.param_groups:
+            if 'step' in group:
+                group['step'] = step
+            for p in group['params']:
+                if p in state and len(state[p]) > 0 and 'step' in state[p]:
+                    if isinstance(state[p]['step'], torch.Tensor):
+                        state[p]['step'] = torch.tensor(step, dtype=torch.float, device=p.device)
```

Comment:
not sure if you want to hardcode fp32 here, perhaps this would be better?
```suggestion
                        state[p]['step'] = torch.tensor(step, dtype=p.dtype, device=p.device)
```

Reply:
state[p]['step'] information holds the step number. I don't think that the step dtype is related to the parameter dtype.
also, torch.optim.adamw uses float (see below):
```
                if len(state) == 0:
                    state['step'] = torch.zeros((1,), dtype=torch.float, device=p.device) \
                        if self.defaults['capturable'] else torch.tensor(0.)
```

Reply:
ah, good catch, yes!

There are variations of AdamW that use lower precision than fp32 and we want to make sure those are supported.

But actually I wonder why is it even fp32 in the first place when it could be `uint` - after all it's step counter which is every increasing natural number.

Reply:
ByteTensor, assuming unsigned-8-bit-byte, will have a max value of 256 - not enough for tracking steps

Reply:
heh, that wasn't a very smart suggestion of mine. Good thing you're being vigilant, Moshe.

I guess `torch.IntTensor` would be the most appropriate then, as half would be 64k max - who knows if someone will want more than 64k steps.

I wonder why is it even converted into a tensor in the first place, since `tensor + int` math works just fine.

Would anything break if we ignore the original dtype and assign a normal Python int regardless?

Reply:
I recall that it was originally `int` and a `group` state, and then changed into tensor and `param` state (in torch or nvidia Adam), which broke DeepSpeed.

Reply:
which is odd, since 
```
python -c "import torch; print(torch.ones(1)+1)"
tensor([2.])
```

unless it was expecting `step` to be a tensor and was doing some accessor call, like `step.dtype`

Reply:
To be clear, the breakage was in checkpoint loading if I recall correctly.

Reply:
I see - it must have been expecting a tensor but got an int instead.

which probably means the checkpoint came from elsewhere? Otherwise if you were to save an int, it should have been fine to load an int back.

I don't think there is any issue with how this is done, it's just odd when float is used for something that is not a float.